components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt

```diff
@@ -65,80 +91,203 @@ class TransactionSignatureServiceImpl @Activate constructor(
         }
     }
 
-    override fun verifySignature(transactionId: SecureHash, signatureWithMetadata: DigitalSignatureAndMetadata) {
+    @Suspendable
+    override fun signBatch(
+        transactions: List<TransactionWithMetadata>,
+        publicKeys: Iterable<PublicKey>
+    ): List<List<DigitalSignatureAndMetadata>> {
+        confirmBatchSigningRequirements(transactions)
+
+        val publicKeysToSigSpecs = getAvailableKeysFor(publicKeys).associateWith { publicKey ->
+            requireNotNull(signatureSpecService.defaultSignatureSpec(publicKey)) {
+                "There are no available signature specs for this public key. ($publicKey ${publicKey.algorithm})"
+            }
+        }
+
+        val hashDigestProvider = transactions.first().getBatchMerkleTreeDigestProvider(merkleTreeProvider)
+        val batchTree = merkleTreeProvider.createTree(transactions.map { it.id.bytes }, hashDigestProvider)
+
+        val batchSignaturesWithMeta = publicKeysToSigSpecs.map { (publicKey, signatureSpec) ->
+            val signatureMetadata =
+                getSignatureMetadata(signatureSpec, getBatchSignatureMetadataSettings(transactions.first()))
+            val signableData = SignableData(batchTree.root, signatureMetadata)
+
+            signingService.sign(
+                serializationService.serialize(signableData).bytes,
+                publicKey,
+                signatureSpec
+            ) to signatureMetadata
+        }
+        return List(transactions.size) {
+            val proof = batchTree.createAuditProof(listOf(it))
+            batchSignaturesWithMeta.map { (signature, signatureMetadata) ->
+                DigitalSignatureAndMetadata(signature, signatureMetadata, proof)
+            }
+        }
+    }
+
+    override fun verifySignature(
+        transaction: TransactionWithMetadata,
+        signatureWithMetadata: DigitalSignatureAndMetadata
+    ) {
         val signatureSpec = checkAndGetSignatureSpec(signatureWithMetadata)
 
-        val signedData = SignableData(transactionId, signatureWithMetadata.metadata)
-        digitalSignatureVerificationService.verify(
+        val proof = signatureWithMetadata.proof
+        val signedHash = if (proof == null) {   // Simple signature
+            transaction.id
+        } else {                                // Batch signature
+
+            confirmSignatureBatchMerkleSettingsMatchWithTransactions(transaction, signatureWithMetadata)
+            confirmHashPrefixesAreDifferent(listOf(transaction))
+
+            require(proof.leaves.filter { transaction.id.bytes contentEquals it.leafData }.size == 1) {
+                "The transaction id should be proven by the proof."
```

Comment:
How about someting like:
```suggestion
                ""The transaction id cannot not be found in the provided Merkle proof for the the batch signature - the signature cannot be verified.""
```

Reply:
Thanks. Applied.